File: Lab(ICE-2202)CPP/lab-09.cpp
#include <iostream>
#include <vector>
using namespace std;

void knapsack(vector<int> &P, vector<int> &W, int C, int n)
{
    vector<vector<int>> dp(n + 1, vector<int>(C + 1, 0));
    vector<vector<bool>> selected(n + 1, vector<bool>(C + 1, false));

    for (int i = 1; i <= n; ++i)
    {
        for (int w = 1; w <= C; ++w)
        {
            if (W[i - 1] <= w)
            {
                dp[i][w] = max(P[i - 1] + dp[i - 1][w - W[i - 1]], dp[i - 1][w]);
                if (P[i - 1] + dp[i - 1][w - W[i - 1]] > dp[i - 1][w])
                {
                    selected[i][w] = true;
                }
            }
            else
            {
                dp[i][w] = dp[i - 1][w];
            }
        }
    }

    cout << "Maximum profit: " << dp[n][C] << endl;
    cout << "Selected items: ";
    int i = n, j = C;
    while (i > 0 && j > 0)
    {
        if (selected[i][j])
        {
            cout << i << " ";
            j -= W[i - 1];
        }
        --i;
    }
    cout << endl;
}

int main()
{
    int n, C;
    cout << "Enter the number of items: ";
    cin >> n;
    cout << "Enter the knapsack capacity: ";
    cin >> C;

    vector<int> P(n), W(n);
    cout << "Enter the profits of items: ";
    for (int i = 0; i < n; ++i)
    {
        cin >> P[i];
    }
    cout << "Enter the weights of items: ";
    for (int i = 0; i < n; ++i)
    {
        cin >> W[i];
    }

    knapsack(P, W, C, n);

    return 0;
}

File: lab-09.py
def knapsack(P, W, C, n):
    # Initialize a 2D array to store the maximum profit for each subproblem
    dp = [[0] * (C + 1) for _ in range(n + 1)]

    # Fill the dp array using dynamic programming
    for i in range(n + 1):
        for w in range(C + 1):
            if i == 0 or w == 0:
                dp[i][w] = 0
            elif W[i - 1] <= w:
                dp[i][w] = max(P[i - 1] + dp[i - 1][w - W[i - 1]], dp[i - 1][w])
            else:
                dp[i][w] = dp[i - 1][w]

    # Trace back to find the items included in the knapsack
    included_items = []
    i, w = n, C
    while i > 0 and w > 0:
        if dp[i][w] != dp[i - 1][w]:
            included_items.append(i - 1)
            w -= W[i - 1]
        i -= 1

    # Return the maximum profit and the list of included items
    return dp[n][C], included_items


# Input section to take the values of P, W, C, and n from the user
P = list(map(int, input("Enter the profits of items separated by space: ").split()))
W = list(map(int, input("Enter the weights of items separated by space: ").split()))
C = int(input("Enter the knapsack capacity: "))
n = int(input("Enter the number of items: "))

max_profit, items_included = knapsack(P, W, C, n)
print("Maximum profit:", max_profit)
print("Items included:", items_included)
